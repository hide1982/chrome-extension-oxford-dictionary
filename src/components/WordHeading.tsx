import React from "react"
import styled from "styled-components"
import { Pronunciation, Result } from "types"
import IconButton from "./IconButton"
import Typography from "./Typography"

interface Props {
  values: {
    word: Result["word"]
    audioFile: Pronunciation["audioFile"]
    phoneticSpelling: Pronunciation["phoneticSpelling"]
  }
}

const Container = styled.div`
  display: flex;
  align-items: center;
  margin: 12px 0;
  padding-bottom: 8px;
  border-bottom: solid 2px ${({ theme }) => theme.borderColor.main};
`

const PhoneticSpelling = styled(Typography)`
  margin-left: 16px;
`

const SpeakerIconButton = styled(IconButton)`
  margin-left: 16px;
`

const playAudio = (audioFile: Pronunciation["audioFile"]) => () => {
  const audio = new Audio(audioFile)
  audio.play()
}

const WordHeading: React.FC<Props> = ({ values }) => {
  const { word, audioFile, phoneticSpelling } = values

  return (
    <Container>
      <Typography size={24} weight="bold">
        {word}
      </Typography>
      <PhoneticSpelling>{phoneticSpelling}</PhoneticSpelling>
      <SpeakerIconButton iconName="speaker" onClick={playAudio(audioFile)} />
    </Container>
  )
}

export default WordHeading
