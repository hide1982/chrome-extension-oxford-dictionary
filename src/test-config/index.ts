import "@testing-library/jest-dom"

import "@/test-config/matchMedia"
