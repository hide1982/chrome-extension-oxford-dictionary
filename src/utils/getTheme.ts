import { DARK_MODE_MEDIA_QUERY } from "@/constants"
import { lightTheme, darkTheme } from "@/styles/theme"
import type { Theme } from "@/styles/theme"

const isDarkMode = (): boolean =>
  window.matchMedia(DARK_MODE_MEDIA_QUERY).matches

const getTheme = (): Theme => {
  const theme = isDarkMode() ? darkTheme : lightTheme

  return theme
}

export default getTheme
