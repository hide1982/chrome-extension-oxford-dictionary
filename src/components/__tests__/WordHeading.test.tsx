import React from "react"
import { render, fireEvent } from "@testing-library/react"
import "@/test-config"

import { ThemeProvider } from "styled-components"
import WordHeading from "@/components/WordHeading"
import getTheme from "@/utils/getTheme"

const wordText = "word"
const phoneticSpellingText = "phoneticSpelling"

const setup = () => {
  const onPlayAudio = jest.fn()
  const utils = render(
    <ThemeProvider theme={getTheme()}>
      <WordHeading
        word={wordText}
        phoneticSpelling={phoneticSpellingText}
        onPlayAudio={onPlayAudio}
      />
    </ThemeProvider>
  )

  const wordHeading = utils.getByTestId("word-heading")
  const speakerButton = utils.getByTestId("word-heading_speaker-button")
  const word = utils.getByTestId("word-heading_word")
  const phoneticSpelling = utils.getByTestId("word-heading_phonetic-spelling")

  return {
    utils,
    fns: {
      onPlayAudio,
    },
    comps: {
      wordHeading,
      speakerButton,
      word,
      phoneticSpelling,
    },
  }
}

describe("WordHeading component", () => {
  test("snapshot", () => {
    const { comps } = setup()

    expect(comps.wordHeading).toMatchSnapshot()
  })

  test("should call onPlayAudio() when click speaker button", () => {
    const { fns, comps } = setup()
    fireEvent.click(comps.speakerButton)

    expect(fns.onPlayAudio).toBeCalled()
  })

  test("should render word", () => {
    const { comps } = setup()

    expect(comps.word.textContent).toBe(wordText)
  })

  test("should render phonetic spelling", () => {
    const { comps } = setup()

    expect(comps.phoneticSpelling.textContent).toBe(phoneticSpellingText)
  })
})
