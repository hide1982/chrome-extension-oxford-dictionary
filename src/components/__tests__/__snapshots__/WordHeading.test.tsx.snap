// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordHeading component snapshot 1`] = `
<div
  class="sc-jSgupP gfSWvj"
  data-testid="word-heading"
>
  <p
    class="sc-eCssSg bQBTac"
    data-testid="word-heading_word"
    font-size="24"
    font-weight="bold"
  >
    word
  </p>
  <p
    class="sc-eCssSg hApimM sc-gKsewC jvEDbY"
    data-testid="word-heading_phonetic-spelling"
    font-size="16"
  >
    phoneticSpelling
  </p>
  <button
    class="sc-dlfnbm borXTz sc-hKgILt bupsiL sc-iBPRYJ cFrrBw"
    data-testid="word-heading_speaker-button"
  >
    <div
      class="sc-bdfBwQ hxYlxV"
      data-testid="icon"
    >
      <svg
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M296.16,54.142L105.073,170.705H0v169.141h105.032l191.299,118.012L296.16,54.142z M94.253,309.753h-64.16V200.798h64.16
     V309.753z M124.346,316.402V194.198l141.743-86.462l0.124,296.186L124.346,316.402z"
        />
        <path
          d="M445.06,97.169l-20.993,21.563c37.299,36.311,57.839,84.873,57.839,136.741c0,51.672-20.408,100.104-57.465,136.375
     l21.049,21.506C488.38,371.374,512,315.304,512,255.471C512,195.413,488.227,139.193,445.06,97.169z"
        />
        <path
          d="M399.602,136.961l-20.342,22.178c26.984,24.75,42.461,59.863,42.461,96.334c0,36.278-15.342,71.265-42.09,95.986
     l20.426,22.099c32.892-30.401,51.757-73.442,51.757-118.087C451.813,210.591,432.782,167.395,399.602,136.961z"
        />
        <path
          d="M353.386,176.586l-18.691,23.585c17.056,13.517,26.838,33.673,26.838,55.302c0,21.448-9.647,41.499-26.467,55.011
     l18.845,23.462c23.968-19.253,37.715-47.855,37.715-78.472C391.627,224.598,377.69,195.845,353.386,176.586z"
        />
      </svg>
    </div>
    <span
      class="sc-gsTCUz jeRPZM"
      data-testid="ripple"
    />
  </button>
</div>
`;
